src/main/kotlin/org/rust/lang/core/psi/RustTokenType.kt

```diff
@@ -1,10 +1,36 @@
 package org.rust.lang.core.psi
 
+import com.intellij.lang.ASTNode
+import com.intellij.lang.DefaultASTFactory
+import com.intellij.openapi.components.service
+import com.intellij.psi.impl.source.tree.LeafPsiElement
 import com.intellij.psi.tree.IElementType
+import com.intellij.psi.tree.ILeafElementType
 import org.rust.lang.RustLanguage
+import org.rust.lang.core.psi.impl.IRustLiteralImplBase
 
-open class RustTokenType(debugName: String) : IElementType(debugName, RustLanguage)
+private val defaultASTFactory by lazy { service<DefaultASTFactory>() }
+
+open class RustTokenType(debugName: String) : IElementType(debugName, RustLanguage), ILeafElementType {
+    override fun createLeafNode(leafText: CharSequence?): ASTNode = LeafPsiElement(this, leafText)
+}
+
+class RustCommentTokenType(debugName: String) : RustTokenType(debugName) {
+    override fun createLeafNode(leafText: CharSequence?): ASTNode = defaultASTFactory.createComment(this, leafText!!)
+}
 
 class RustKeywordTokenType(debugName: String) : RustTokenType(debugName)
 
-class RustLiteralTokenType(debugName: String) : RustTokenType(debugName)
+class RustLiteralTokenType(
+    debugName: String,
+    private val implClass: Class<out IRustLiteralImplBase>
+) : RustTokenType(debugName) {
+    override fun createLeafNode(leafText: CharSequence?): ASTNode {
+        try {
+            val constructor = implClass.getConstructor(IElementType::class.java, CharSequence::class.java)
+            return constructor.newInstance(this, leafText!!)
+        } catch(e: Exception) {
+            throw RuntimeException(e)
+        }
```

Comment:
Can you pass a function instead of a class in the constructor?

```
RustTokenType INTEGER_LITERAL = new RustLiteralTokenType("<INTEGER>", (tokenType, text) -> new RustNumericLiteralImpl(tokenType, text));
```

Or even better


```
RustTokenType INTEGER_LITERAL = new RustLiteralTokenType("<INTEGER>", RustNumericLiteralImpl::create);
```